### ISSUE TITLE
Suggestion: don't log stacktrace in case of a `CancellationException`

### ISSUE THREAD

**Issue by Stephan202** [ben-manes/caffeine#597]
Caffeine logs a warning including stack trace if a `Future` completes exceptionally ([(1)](https://github.com/ben-manes/caffeine/blob/b85c8ec80e867ba7c6d17aa235f601f5d90ad00f/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L1290-L1297), [(2)](https://github.com/ben-manes/caffeine/blob/0dcb4197d6df9884a8bd678b3583b2eb5cfe3379/caffeine/src/main/java/com/github/benmanes/caffeine/cache/LocalLoadingCache.java#L130-L137), [(3)](https://github.com/ben-manes/caffeine/blob/c42ab535261715f654f9c116ccb984ba3c69f6f4/caffeine/src/main/java/com/github/benmanes/caffeine/cache/LocalAsyncLoadingCache.java#L280-L287), [(4)](https://github.com/ben-manes/caffeine/blob/b85c8ec80e867ba7c6d17aa235f601f5d90ad00f/caffeine/src/main/java/com/github/benmanes/caffeine/cache/LocalAsyncCache.java#L193-L202), maybe elsewhere). This is important for flows in which the exception isn't bubbled up, as it may otherwise hide application errors.

However, this code also logs a full stack trace if a `Future` is simply cancelled. The log line then states than a `java.util.concurrent.CancellationException` was thrown, while strictly speaking this exception was only created inside `CompletableFuture#cancel(boolean)`.

My change request is this: would it be possible to identify `CancellationException`s and log those differently, e.g. as:
```java
if (error instanceof CancellationException) {
  logger.log(Level.INFO, "Refresh task was cancelled");
} else {
  logger.log(Level.WARNING, "Exception thrown during refresh", error);
}
```

(This example is for the `BoundedLocalCache` case; I also changed `WARNING` to `INFO`, but have no strong opinion on that, actually.)

Context: we have a custom cache loader, and if its background bulk refresh logic can't keep up, it cancels the oldest refresh requests. When this happens a (very) large number of identical and rather uninformative stack traces is logged. In extreme cases this  also causes our log agent to have trouble keeping up.

**Comment by ben-manes** [comment]
I think that's a good observation that cancellation should be special cased and is unlike the logging's intent. Do you think it should be logged at all? Either way seems reasonable to me, but I'd lean towards thinking cancellation is an explicit choice and expected if done, whereas most other exceptions are a surprise that the application is likely to silently swallow.

**Comment by Stephan202** [comment]
> Do you think it should be logged at all?

I'd also be fine with not logging at all. Alternatively `Level.TRACE` could have the same effect in most setups. From our side either work be just fine. :+1:

**Comment by Stephan202** [comment]
Tnx @ben-manes!

**Comment by ben-manes** [comment]
rolling this back from master as the CI is unhappy, but seems unrelated. If it fails only on the v3.dev branch then I'll have something to go off of, but I'm afraid it might be stuck failing on a prior green commit.

Unfortunately soft reference caching is flaky to test because we have to coerce the GC to maybe discard the entries. We do this using `-XX:SoftRefLRUPolicyMSPerMB=0` and `-XX:+UseParallelGC` with Guava's `GcFinalization.awaitFullGc()`. That works most of the time and locally, but the CI is more finicky. That could be something minor like a new JDK patch release that tweaks the GC to break this, a VM's noisy neighbor causing fewer GCs due to cpu pressure, etc. It mostly. works but sometimes has bad days like now.

**Comment by ben-manes** [comment]
I'm not sure why, but debugging this and the [slf4j-test](https://github.com/valfirst/slf4j-test) logger causes the issues. There is nothing special about that logger, nothing is logged which would explain it, and it is cleared between tests. Either the logging tests should be dropped or the logger isolated so that it isn't used by the reference tests.

**Comment by ben-manes** [comment]
disabling the test loggers by default seems to have stabilized it.

**Comment by ben-manes** [comment]
Sorry for dragging my feet here. This is released in v2.9.3 and v3.0.5.
